emacs/config.org
#+TITLE: Emacs configuration
#+DESCRIPTION: Emacs configuration is written in orgmode. Code is directly written to the files using org-babel-tangle without the need to start orgmode at startup.
#+PROPERTY: header-args :lexical t :tangle "init.el" :mkdirp "lisp"
#+STARTUP: showeverything  hidestars

* Early-init
=Early-init= made for speed up Emacs.

#+BEGIN_SRC elisp :tangle "early-init.el"
;; early-init.el -*- lexical-binding: t; -*-

;; Increase 'gc-cons-threshold' to 100MB for lsp heavy buffers.
(setq gc-cons-threshold 100000000)

;; Handle large chunks of process output, improving peroformence of external processes (like LSP servers)
(setq read-process-output-max (* 1024 1024)) ;; 1 MB

(setq load-path-filter-function #'load-path-filter-cache-directory-files)

(setq corfu-debug t)
(setq eglot-events-buffer-size 2000000)  ; Increase for debugging

;; Tree-sitter performence enchancement
(setenv "LSP_USE_PLISTS" "true")
(setq lsp-use-plists t)

;; Disable "file-name-handler-alist" than enable it later for speed.
(defvar startup/file-name-handler-alist file-name-handler-alist)
(setq file-name-handler-alist nil)
(add-hook 'emacs-startup-hook
          (lambda ()
            (setq file-name-handler-alist startup/file-name-handler-alist)
            (makunbound 'startup/file-name-handler-alist)))

;; Speed up package loading
(setq package-quickstart t)

;; Loads packages only when needed
(setq use-package-always-defer t)
#+END_SRC

** Disable menu and scrollbar
#+BEGIN_SRC elisp :tangle "early-init.el"
(tool-bar-mode -1)
(menu-bar-mode -1)
(tooltip-mode -1)
(scroll-bar-mode -1)
(horizontal-scroll-bar-mode -1)
(global-tab-line-mode -1) ; tab line
(tab-bar-mode -1) ; tabs
(global-visual-line-mode t) ; don't let the text to go horizontal.
(global-hl-line-mode t) ; highlight the line you're on.
(global-display-line-numbers-mode t) ; Slow down with big files.
(electric-pair-mode t) ;; auto brackets.
(electric-indent-mode -1)
                                        ;(blink-cursor-mode -1)
                                        ;(desktop-save-mode 1)
#+END_SRC

** Remeber last place you were in file.
#+BEGIN_SRC elisp :tangle "early-init.el"
(save-place-mode 1)
#+END_SRC

** Auto revert buffer for changed files.
#+BEGIN_SRC elisp :tangle "early-init.el"
(global-auto-revert-mode t)
#+END_SRC

** Auto revert dired buffer

#+BEGIN_SRC elisp :tangle "early-init.el"
(setq global-auto-revert-non-file-buffers t)
#+END_SRC
** Open Emacs in full screen always.

#+BEGIN_SRC elisp :tangle "early-init.el"
(add-to-list 'default-frame-alist '(fullscreen . maximized))
#+END_SRC

** Preventing flickering.
#+begin_src elisp :tangle "early-init.el"
(add-to-list 'default-frame-alist '(inhibit-double-buffering . t))
#+end_src

** Show parenthesis.
#+BEGIN_SRC elisp :tangle "early-init.el"
(setq show-paren-style 'expression)
(show-paren-mode 1)
(set-face-background 'show-paren-match (face-background 'default))
(set-face-foreground 'show-paren-match "#def")
(custom-set-faces
 '(show-paren-match ((t (:weight extra-bold)))))
#+END_SRC

** UTF
#+BEGIN_SRC elisp :tangle "early-init.el"
(prefer-coding-system 'utf-8)
(set-default-coding-systems 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
#+END_SRC

** Major settings
#+BEGIN_SRC elisp :tangle "early-init.el"
(setq
 display-line-numbers-width nil
 package-enable-at-startup nil ; don't auto-initialize!
 package-enable-imenu-support t
 package--init-file-ensured t ; don't add that `custom-set-variables' block to init
 package-archives nil
 load-prefer-newer t ; prefer newer package files.
 frame-resize-pixelwise t
 visible-bell 1 ; Alarm bell (flash the bar)

 ;; Tab behaviour
 completion-cycle-threshold 3 ;; TAB cycle if there are only few candidates
 tab-always-indent 'complete ;; Enable indentation+completion using the TAB key.
 tab-line-new-button-show nil
 echo-keystrokes 0.01 ;; decrease the echoing of the keystrokes.

 ;; Tab line
 tab-line-close-button-show nil
 tab-line-separator " "
 line-move-visual nil
 site-run-file nil ; unset SRF
 ;; scroll
 pixel-scroll-precision-mode t
 pixel-scroll-precision-use-momentum t
 scroll-conservatively 101 ;; scroll line by line.

 use-file-dialog nil
 mode-line-format nil ; don't want a mode line while loading init
 create-lockfiles nil ; disable lockfiles
 make-backup-files nil ; disable backup file
 backup-directory-alist '((".*" . "~/.local/share/Trash/files"))

 ;; Disable auto save alltogether.
 auto-save-list-file-prefix nil ; disable auto-save
 auto-save-default nil
 auto-save-mode nil
 delete-auto-save-files t ; delete auto save files

 auto-mode-case-fold nil ; use case-sensitive `auto-mode-alist' for performance
 default-input-method nil
 utf-translate-cjk-mode nil ; disable CJK coding/encoding (Chinese/Japanese/Korean characters)
 initial-scratch-message nil ; empty the initial *scratch* buffer.
 message-log-max nil
 command-line-x-option-alist nil ; remove irreleant command line options for faster startup
 use-short-answers t ; y/n for yes/no
 vc-follow-symlinks t ; Do not ask about symlink following
 use-dialog-box nil
 inhibit-default-init t
 use-file-dialog nil
 redisplay-dont-pause t ; improve display engine

 ;; Suppress the startup
                                        ;inhibit-splash-screen t
                                        ;inhibit-startup-screen t		; do not show the startup message
                                        ;inhibit-startup-message t      ; reduce noise at startup
                                        ;inhibit-startup-buffer-menu t  ; stop `list-buffers' from showing when opening multiple files

 frame-inhibit-implied-resize t ; do not resize the frame at this early stage
 ffap-machine-p-known 'reject   ; don't ping things that look like domain names

 ;; Improve performance
 inhibit-compacting-font-caches t ; Inhibit frame resizing for performance
 read-process-output-max (* 1024 1024) ; Increase how much is read from processes in a single chunk.
 fast-but-imprecise-scrolling t ; more performant rapid scrolling over unfontified regions
 scroll-conservatively 101
 scroll-margin 0
 scroll-preserve-screen-position t

 redisplay-skip-fontification-on-input t ; Inhibits it for better scrolling performance.
 idle-update-delay 1.0 ; slow down UI updates down
 select-active-regions 'only ; Emacs hangs when large selections contain mixed line endings
 ad-redefinition-action 'accept ; disable warnings from legacy advice system
 initial-major-mode 'fundamental-mode
 confirm-kill-emacs 'y-or-n-p ; confirm before exiting emacs
 enable-local-variables :safe ; host security
 column-number-mode t
 org-indent-mode nil
 display-line-numbers-type 'relative
 custom-file (make-temp-file "emacs-custom-")
 enable-recursive-minibuffers t                ; Use the minibuffer whilst in the minibuffer
 completion-cycle-threshold 1                  ; TAB cycles candidates
 completions-detailed t                        ; Show annotations
 tab-always-indent 'complete                   ; When I hit TAB, try to complete, otherwise, indent
 completion-styles '(basic initials substring) ; Different styles to match input to candidates

 completion-auto-help 'always                  ; Open completion always; `lazy' another option
 completions-max-height 20                     ; This is arbitrary
 completions-detailed t
 completions-format 'one-column
 completions-group t
 completion-auto-select 'second-tab            ; Much more eager
 x-stretch-cursor nil


 ;; Vertical Scroll
 scroll-step 1
 scroll-margin 1
 scroll-conservatively 101
 scroll-up-aggressively 0.01
 scroll-down-aggressively 0.01
 auto-window-vscroll nil
 fast-but-imprecise-scrolling nil
 mouse-wheel-scroll-amount '(1 ((shift) . 1))
 mouse-wheel-progressive-speed nil
 ;; Horizontal Scroll
 hscroll-margin 2
 hscroll-step 1
 ;; Emacs spends too much effort recentering the screen if you scroll the
 ;; cursor more than N lines past window edges (where N is the settings of
 ;; `scroll-conservatively'). This is especially slow in larger files
 ;; during large-scale scrolling commands. If kept over 100, the window is
 ;; never automatically recentered. The default (0) triggers this too
 ;; aggressively, so I've set it to 10 to recenter if scrolling too far
 ;; off-screen.
 scroll-conservatively 101
 scroll-preserve-screen-position t
 scroll-margin 2
 scroll-preserve-screen-position t
 ;; Reduce cursor lag by a tiny bit by not auto-adjusting `window-vscroll'
 ;; for tall lines.
 auto-window-vscroll nil
 ;; mouse
 mouse-wheel-scroll-amount '(2 ((shift) . hscroll))
 mouse-wheel-scroll-amount-horizontal 2

 confirm-nonexistent-file-or-buffer nil

                                        ;  (setq-default isearch-lazy-count t)
 enable-recursive-minibuffers t
 kill-ring-max 100

                                        ; frame-title-format "E M A C S"

                                        ; browse-url-browser-function 'browse-url-xdg-open

                                        ; custom-safe-themes t

 native-comp-async-report-warnings-errors nil

 ;; Prevent unwanted runtime builds in gccemacs (native-comp); packages are
 ;; compiled ahead-of-time when they are installed and site files are compiled
 ;; when gccemacs is installed.
 comp-deferred-compilation nil

 ;; Compile all sites-lisp on demand.
 native-comp-jit-compilation t

 ;; Keep the eln cache clean.
 native-compile-prune-cache t

 ;; Solve slow icon rendering
 inhibit-compacting-font-caches t

 ;; Enable ibuffer
 ibuffer-expert t

 display-buffer-alist nil

 select-enable-clipboard t ;; Copy and Paste outside of Emacs
 )

(defalias 'yes-or-no-p 'y-or-n-p) ; yes or no to y or n
                                        ;  (add-hook 'prog--hook 'display-line-numbers-mode) ; Only use line-numbers in major modes
                                        ;  (add-hook 'text-mode-hook 'display-line-numbers-mode)
(windmove-default-keybindings)

;; Improve memory
(setq-default history-length 1000)
(setq-default prescient-history-length 1000)
#+END_SRC

** Fonts
#+BEGIN_SRC elisp :tangle "early-init.el"
;; Set base faces for all users
(custom-set-faces
 ;; Default font for all text
                                        ; '(default ((t (:family "JetBrains Mono" :height 90))))
                                        ;'(fixed-pitch ((t (:family "JetBrains Mono" :height 80))))

 ;; Current line number
 '(line-number-current-line ((t (:foreground "yellow" :inherit line-number))))
 '(mode-line ((t (:family "JetBrains Mono" :weight Bold))))

 ;; Comments italic
 ;; '(font-lock-comment-face ((t (:family "JetBrains Mono" :slant italic))))
 ;; ;;   ;; Keywords, functions, strings, etc. italic with no color change
 ;; '(font-lock-keyword-face ((t (:family "JetBrains Mono" :slant italic))))
 ;; '(font-lock-function-name-face ((t (:family "JetBrains Mono":slant italic))))
 ;; '(font-lock-string-face ((t (:family "JetBrains Mono" :slant italic))))
 ;; '(font-lock-variable-name-face ((t (:family "JetBrains Mono":weight bold))))
 ;; ;;   '(font-lock-constant-face ((t (:family "JetBrains Mono" :slant italic))))
 ;; ;;   '(font-lock-type-face ((t (:family "JetBrains Mono" :slant italic))))
 ;; ;;   '(font-lock-builtin-face ((t (:family "JetBrains Mono" :slant italic))))

 ;; `(font-lock-builtin-face ((t (:family "JetBrains Mono" :weight bold :slant italic))))
 ;;     `(font-lock-comment-face ((t (:family "JetBrains Mono" :slant italic))))
 ;;     `(font-lock-constant-face ((t (:family "JetBrains Mono" :weight bold))))
 ;;     `(font-lock-function-name-face ((t (:family "JetBrains Mono" :weight bold))))
 ;;     `(font-lock-keyword-face ((t (:family "JetBrains Mono" :weight bold :slant italic))))
 ;;     `(font-lock-string-face ((t (:family "JetBrains Mono" :slant italic))))
 ;;     `(font-lock-type-face ((t (:family "JetBrains Mono" :weight bold))))
 ;;     `(font-lock-variable-name-face ((t (:family "JetBrains Mono" :weight bold))))
 ;;     `(font-lock-warning-face ((t (:family "JetBrains Mono" :weight bold :underline t))))

 )

(when (string= (system-name) "WarMachine")
  (custom-set-faces
   '(default ((t (:family "JetBrains Mono" :height 100))))
   '(variable-pitch ((t (:family "Linux Libertine" :height 80))))
   '(fixed-pitch ((t (:family "JetBrains Mono" :height 90))))))

(when (string= (system-name) "Tiffany")
  (setq doom-modeline-height 20)
  (setq display-battery-mode nil)
  (custom-set-faces
   '(default ((t (:family "JetBrains Mono" :height 90))))
   '(variable-pitch ((t (:family "Linux Libertine" :height 80))))
   '(fixed-pitch ((t (:family "JetBrains Mono" :height 80))))))
#+end_src

* Package Management

** Straight
#+BEGIN_SRC elisp
;; init.el -*- lexical-binding: t; -*-
(defvar elpaca-installer-version 0.11)
(defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
(defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
(defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
(defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                              :ref nil :depth 1 :inherit ignore
                              :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                              :build (:not elpaca--activate-package)))
(let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
       (build (expand-file-name "elpaca/" elpaca-builds-directory))
       (order (cdr elpaca-order))
       (default-directory repo))
  (add-to-list 'load-path (if (file-exists-p build) build repo))
  (unless (file-exists-p repo)
    (make-directory repo t)
    (when (<= emacs-major-version 28) (require 'subr-x))
    (condition-case-unless-debug err
        (if-let* ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                  ((zerop (apply #'call-process `("git" nil ,buffer t "clone"
                                                  ,@(when-let* ((depth (plist-get order :depth)))
                                                      (list (format "--depth=%d" depth) "--no-single-branch"))
                                                  ,(plist-get order :repo) ,repo))))
                  ((zerop (call-process "git" nil buffer t "checkout"
                                        (or (plist-get order :ref) "--"))))
                  (emacs (concat invocation-directory invocation-name))
                  ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                        "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                  ((require 'elpaca))
                  ((elpaca-generate-autoloads "elpaca" repo)))
            (progn (message "%s" (buffer-string)) (kill-buffer buffer))
          (error "%s" (with-current-buffer buffer (buffer-string))))
      ((error) (warn "%s" err) (delete-directory repo 'recursive))))
  (unless (require 'elpaca-autoloads nil t)
    (require 'elpaca)
    (elpaca-generate-autoloads "elpaca" repo)
    (let ((load-source-file-function nil)) (load "./elpaca-autoloads"))))
(add-hook 'after-init-hook #'elpaca-process-queues)
(elpaca `(,@elpaca-order))

;; Compile everything.
;;(setq straight-byte-compile t) ;; ensure byte compilation is enabled
;;(native-compile-async (expand-file-name "straight/build" straight-base-dir) 'recursively)
(setq package-enable-at-startup nil)
#+END_SRC

** Use-package
#+BEGIN_SRC elisp
(elpaca elpaca-use-package
  ;; Enable use-package :ensure support for Elpaca.
  (elpaca-use-package-mode))

(setq elpaca-use-package-by-default t)

(require 'use-package)

(setq use-package-compute-statistics t
      package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("org" . "https://orgmode.org/elpa/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))
#+END_SRC

** Maximum native eln speed.
Look for native-compile-async using variable "C-h v =native-comp-eln-load-path="
#+BEGIN_SRC elisp

(when (string= (system-name) "WarMachine")
  (setq native-comp-speed 3)
  (setq native-comp-compiler-options '("-march=znver2" "-Ofast" "-g0" "-fno-finite-math-only" "-fgraphite-identity" "-floop-nest-optimize" "-fdevirtualize-at-ltrans" "-fipa-pta" "-fno-semantic-interposition" "-flto=auto" "-fuse-linker-plugin"))
  (setq native-comp-driver-options '("-march=znver2" "-Ofast" "-g0" "-fno-finite-math-only" "-fgraphite-identity" "-floop-nest-optimize" "-fdevirtualize-at-ltrans" "-fipa-pta" "-fno-semantic-interposition" "-flto=auto" "-fuse-linker-plugin"))
  (native-compile-async "/usr/local/lib/emacs/31.0.50/native-lisp" 'recursively))

(when (string= (system-name) "Tiffany")
  (setq native-comp-speed 3)
  (setq native-comp-compiler-options '("-march=znver3" "-Ofast" "-g0" "-fno-finite-math-only" "-fgraphite-identity" "-floop-nest-optimize" "-fdevirtualize-at-ltrans" "-fipa-pta" "-fno-semantic-interposition" "-flto=auto" "-fuse-linker-plugin"))
  (setq native-comp-driver-options '("-march=znver3" "-Ofast" "-g0" "-fno-finite-math-only" "-fgraphite-identity" "-floop-nest-optimize" "-fdevirtualize-at-ltrans" "-fipa-pta" "-fno-semantic-interposition" "-flto=auto" "-fuse-linker-plugin"))
  (native-compile-async "/usr/lib/emacs/30.2/native-lisp" 'recursively))
#+END_SRC

* General settings

** Emacs
#+BEGIN_SRC elisp
;; Global tab width and use spaces
(use-package emacs
  :ensure nil
  :defer t
  :bind
  (("<C-wheel-down>" . text-scale-increase)
   ("<C-wheel-up>" . text-scale-decrease)
   ("<f11>" . 'my/toggle-mode-line)        ;; toggle modeline
   ("<f12>" . 'my/zen-mode)        ;; toggle olivetti (zen mode)
   ("C-+" . text-scale-increase)
   ("C--" . text-scale-decrease)
   ("C-." . avy-goto-char)
   ;; C-c
   ("C-c C-u" . package-upgrade-all)
   ("C-c R" . restart-emacs)              ;; restart emacs
   ("C-c T" . vterm)
   ("C-c b" . nyan-mode)
   ("C-c c" . compile)
   ("C-c g" . gdb)
   ("C-c i" . 'my/indent-whole-buffer)
   ("C-c k" . 'my/kill-all-buffers)        ;; kill all buffers
   ("C-c n" . neotree-toggle)
   ("C-c p" . dmenu)
   ("C-c q" . 'my/visit-qtile)             ;; visit qtile config
   ("C-c r" . recentf)                    ;; recent files list
   ("C-c t" . 'vterm-toggle)
   ("C-c v" . view-mode)
   ("C-c w h" . 'my/highlight-word)
   ("C-c w l" . 'my/copy-whole-line)
   ("C-c w w" . 'my/kill-whole-word)
   ("C-c y" . yas-expand)
   ;; C-x
   ("C-x 2" . 'my/split-and-follow-horizontally)
   ("C-x 3" . 'my/split-and-follow-vertically)
   ("C-x B" . infu-bionic-reading-buffer)
   ("C-x C-k" . kill-buffer)              ;; kill buffer
   ("C-x b" . consult-buffer)
   ("C-z" . repeat)
   ("M-y" . popup-kill-ring)
   ("C-c e" . 'my/visit-init)              ;; visit init.el
   ;; Project
                                        ;   ("C-c p p" . project-switch-project)
                                        ;  ("C-c p f" . project-find-file)

   ;; ;; Consult
   ;; ([remap bookmark-jump] . consult-bookmark)
   ;; ([remap goto-line] . consult-goto-line)
   ;; ([remap imenu] . consult-imenu)
   ;; ([remap project-switch-to-buffer] . consult-project-buffer)
   ;; ([remap repeat-complex-command] . consult-complex-command)
   ;; ([remap switch-to-buffer] . consult-buffer)
   ;; ([remap yank-pop] . consult-yank-pop)

   ("C-c M-x" . consult-mode-command)
                                        ;   ("C-c h" . consult-history)
   ("C-c k" . consult-kmacro)
   ("C-c m" . consult-man)
   ("C-c i" . consult-info)
   ("C-x M-:" . consult-complex-command)
   ("C-x b" . consult-buffer)
   ("C-x 4 b" . consult-buffer-other-window)
   ("C-x 5 b" . consult-buffer-other-frame)
   ("C-x t b" . consult-buffer-other-tab)
   ("C-x r b" . consult-bookmark)
   ("C-x p b" . consult-project-buffer)
   ("M-#" . consult-register-load)
   ("M-'" . consult-register-store)
   ("C-M-#" . consult-register)
   ("M-y" . consult-yank-pop)
   ;; M-g prefix (goto-map)
   ("M-g e" . consult-compile-error)
   ("M-g f" . consult-flymake)

qtile/config.py
import os
import subprocess

from libqtile import bar, hook, layout, qtile, widget
from libqtile.config import (Click, Drag, DropDown, Group, Key, Match,
                             ScratchPad, Screen)
from libqtile.dgroups import simple_key_binder
from libqtile.lazy import lazy

dgroups_key_binder = simple_key_binder("mod4")
mod = "mod4"
terminal = "alacritty"
browser = "librewolf --profilemanager"
dmenu = 'dmenu_run -nb "#000000" -nf "#ffffff" -sb "#ffffff" -sf "#000000"'
filemanager = "thunar"
ytfzf = "ytfzf -m -D"

# Colors
bfg = "#ffffff"  # bar foreground color
fg = "#ffffff"  # task switcher color
bg = "#000000"  # bar background color
f_color = "#ED7014"  # focus color

keys = [
    Key([mod], "space", lazy.layout.next(), desc="Move window focus to other window"),
    Key(
        [mod, "shift"], "h", lazy.layout.shuffle_left(), desc="Move window to the left"
    ),
    Key(
        [mod, "shift"],
        "l",
        lazy.layout.shuffle_right(),
        desc="Move window to the right",
    ),
    Key([mod, "shift"], "j", lazy.layout.shuffle_down(), desc="Move window down"),
    Key([mod, "shift"], "k", lazy.layout.shuffle_up(), desc="Move window up"),
    Key([mod], "h", lazy.layout.grow_left(), desc="Grow window to the left"),
    Key([mod], "l", lazy.layout.grow_right(), desc="Grow window to the left"),
    Key([mod], "j", lazy.layout.grow_down(), desc="Grow window up"),
    Key([mod], "k", lazy.layout.grow_up(), desc="Grow window up"),
    Key([mod], "r", lazy.layout.normalize(), desc="Reset all window sizes"),
    Key([], "XF86MonBrightnessUp", lazy.spawn("brightnessctl s 5%+")),
    Key([], "XF86MonBrightnessDown", lazy.spawn("brightnessctl s 5%-")),
    Key(
        [mod, "shift"],
        "Return",
        lazy.layout.toggle_split(),
        desc="Toggle between split and unsplit sides of stack",
    ),
    Key([mod, "shift"], "b", lazy.hide_show_bar("top"), desc="Hide Bar"),
    # Null Keys
    Key(["control"], "q", lazy.spawn(), desc="Reset all window sizes"),
    # Applications
    Key(
        [mod],
        "x",
        lazy.spawn(terminal),
    ),
    Key(
        [mod],
        "b",
        lazy.spawn(browser),
    ),
    Key(
        [mod],
        "p",
        lazy.spawn(dmenu),
    ),
    Key(
        [mod],
        "f",
        lazy.spawn(filemanager),
    ),
    Key(
        [mod],
        "q",
        lazy.to_screen(1),
    ),
    Key(
        [mod],
        "w",
        lazy.to_screen(0),
    ),
    Key(
        [mod],
        "y",
        lazy.spawn(ytfzf),
    ),
    # Toggle between different layouts as defined below
    Key([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"),
    Key([mod, "shift"], "c", lazy.window.kill(), desc="Kill focused window"),
    Key([mod, "shift"], "r", lazy.reload_config(), desc="Reload the config"),
    Key(
        [mod, "shift"],
        "q",
        lazy.spawn("./.config/qtile/scripts/shutdown.sh"),
    ),
    # Sound
    Key([], "XF86AudioMute", lazy.spawn("amixer -q set Master toggle")),
    Key([], "XF86AudioLowerVolume", lazy.spawn("amixer -c 0 sset Master 1- unmute")),
    Key([], "XF86AudioRaiseVolume", lazy.spawn("amixer -c 0 sset Master 1+ unmute")),
    Key(["control"], "9", lazy.spawn("amixer -q set Master 5%-")),
    Key(["control"], "0", lazy.spawn("amixer -q set Master 1%+")),
    Key(["control"], "m", lazy.spawn("pactl set-sink-mute 0 toggle")),
    Key(["control", "shift"], "1", lazy.spawn("xrandr --output eDP --off")),
    Key(
        ["control", "shift"],
        "2",
        lazy.spawn("xrandr --output eDP --auto --left-of HDMI-A-0"),
    ),
    Key([mod, "shift"], "f", lazy.window.toggle_fullscreen()),
]

groups = [
    Group("一"),
    Group("二"),
    Group("三"),
    Group("四"),
    Group("五"),
    Group("六"),
    Group("七"),
    Group("八"),
    Group("九"),
    Group("十"),
]

# ScratchPad
groups.append(
    ScratchPad(
        "scratchpad",
        [
            DropDown(
                "term",
                "alacritty",
                width=0.5,
                height=0.55,
                x=0.26,
                y=0.1,
                opacity=1,
                on_focus_lost_hide=True,
                warp_pointer=True,
            ),
            DropDown(
                "nvim",
                "neovide",
                width=0.5,
                height=0.55,
                x=0.26,
                y=0.1,
                opacity=1,
                on_focus_lost_hide=True,
            ),
            DropDown(
                "ranger",
                "alacritty -e ranger",
                width=0.5,
                height=0.55,
                x=0.26,
                y=0.1,
                opacity=1,
                on_focus_lost_hide=True,
            ),
            DropDown(
                "emacs",
                "emacs",
                width=0.5,
                height=0.55,
                x=0.26,
                y=0.1,
                opacity=1,
                on_focus_lost_hide=True,
            ),
        ],
    )
)

keys.extend(
    [
        Key([mod], "t", lazy.group["scratchpad"].dropdown_toggle("term")),
        Key([mod], "n", lazy.group["scratchpad"].dropdown_toggle("nvim")),
        Key([mod], "r", lazy.group["scratchpad"].dropdown_toggle("ranger")),
        Key([mod], "e", lazy.group["scratchpad"].dropdown_toggle("emacs")),
    ]
)

layout_theme = {
    "border_width": 3,
    "margin": 0,
    "border_focus": f_color,
    "grow_amount": 2,
}

layouts = [
    layout.Columns(**layout_theme),
    # layout.Max(),
    # Try more layouts by unleashing below layouts.
    # layout.Stack(num_stacks=2),
    # layout.Bsp(),
    # layout.Matrix(),
    # layout.MonadTall(),
    # layout.MonadWide(),
    # layout.RatioTile(),
    # layout.Tile(),
    # layout.TreeTab(),
    # layout.VerticalTile(),
    # layout.Zoomy(),
]


widget_defaults = dict(
    font="Roboto Mono Medium",
    # font="JetBrainsMono NFM Medium",
    fontsize=10,
    padding=5,
    background=bg,
    foreground=fg,
)

extension_defaults = widget_defaults.copy()


def init_widgets_list():
    widgets_list = [
        widget.Systray(
            fmt="[{}]",
        ),
        widget.WindowName(
            max_chars=30,
            for_current_screen="False",
            format="{state}{name}",
            markup="True",
            foreground=bfg,
            padding=5,
        ),
        widget.Spacer(),
        widget.GroupBox(
            this_current_screen_border="#2677AB",
            highlight_method="line",
            center_aligned="True",
            active="#ffffff",
            inactive="#404040",
            disable_drag="False",
            padding=5,
            borderwidth=5,
        ),
        widget.Spacer(),
        widget.GenPollCommand(
            update_interval=100,
            cmd="~/.config/qtile/scripts/emacs-daemon.sh",
            shell=True,
            fmt="[{}]",
        ),
        # widget.Spacer(),
        widget.Pomodoro(
            color_break={"ffff00"},
            notification_on=True,
            length_pomodori=90,
            length_long_break=10,
            lenght_short_break=5,
            color_active="ffffff",
            color_inactive=bfg,
            prefix_inactive="Pomodoro",
            prefix_break="Break",
            prefix_long_break="LB",
            prefix_paused="Paused",
            fmt="[{}]",
            padding=5,
        ),
        widget.CheckUpdates(
            distro="Arch",
            update_interval="62800",
            no_update_string="Updated",
            fmt="[{}]",
            display_format="󰚰: {updates}",
            scroll_hide="False",
        ),
        widget.CPU(
            format="{load_percent}%",
            foreground=bfg,
            padding=5,
            fmt="[󰻠: {}]",
            update_interval=4,
        ),
        widget.Memory(
            format="{MemUsed: .0f}",
            fmt="[:{}]",
            foreground=bfg,
            padding=5,
            update_interval=4,
        ),
        # widget.Spacer(),
        widget.ThermalSensor(
            tag_sensor="Tctl",
            foreground=bfg,
            fmt="[󱤋: {}]",
            padding=5,
            threshold=80,
            update_interval=4,
        ),
        widget.Volume(
            foreground=bfg,
            fmt="[: {}]",
            threshold=80,
            padding=5,
        ),
        widget.Battery(
            format="{percent:2.0%} {hour:d}:{min:02d}",
            notify_below=30,  # Show a notification if battery is below 10%
            notification_timeout=10,  # Notification is visible for 10s
        ),
        widget.Clock(
            format="%a %I:%M%p %Y-%m-%d",
            fmt=" [{}]",
            foreground=bfg,
            padding=5,
            markup="True",
        ),
    ]
    return widgets_list


def init_widgets_screen1():
    widgets_screen1 = init_widgets_list()
    del widgets_screen1[
        9:10
    ]  # Slicing removes unwanted widgets (systray) on Monitors 1,3
    return widgets_screen1


def init_widgets_screen2():
    widgets_screen2 = init_widgets_list()
    return widgets_screen2  # Monitor 2 will display all widgets in widgets_list


def init_screens():
    return [
        Screen(top=bar.Bar(widgets=init_widgets_screen1(), opacity=1.0, size=20)),
        Screen(top=bar.Bar(widgets=init_widgets_screen2(), opacity=1.0, size=20)),
        Screen(top=bar.Bar(widgets=init_widgets_screen1(), opacity=1.0, size=20)),
    ]


if __name__ in ["config", "__main__"]:
    screens = init_screens()
    widgets_list = init_widgets_list()
    widgets_screen1 = init_widgets_screen1()
    widgets_screen2 = init_widgets_screen2()


def window_to_prev_group(qtile):
    i = qtile.groups.index(qtile.currentGroup)
    qtile.currentWindow.togroup(qtile.groups[i - 1].name)


def window_to_next_group(qtile):
    if qtile.currentWindow is not None:
        i = qtile.groups.index(qtile.currentGroup)
        qtile.currentWindow.togroup(qtile.groups[i + 1].name)


def window_to_previous_screen(qtile):
    i = qtile.screens.index(qtile.current_screen)
    if i != 0:
        group = qtile.screens[i - 1].group.name
        qtile.current_window.togroup(group)


def window_to_next_screen(qtile):
    i = qtile.screens.index(qtile.current_screen)
    if i + 1 != len(qtile.screens):
        group = qtile.screens[i + 1].group.name
        qtile.current_window.togroup(group)


def switch_screens(qtile):
    i = qtile.screens.index(qtile.current_screen)
    group = qtile.screens[i - 1].group
    qtile.current_screen.set_group(group)


mouse = [
    Drag(
        [mod],
        "Button1",
        lazy.window.set_position_floating(),
        start=lazy.window.get_position(),
    ),
    Drag(
        [mod], "Button3", lazy.window.set_size_floating(), start=lazy.window.get_size()
    ),
    Click([mod], "Button2", lazy.window.bring_to_front()),
]

groups_app_rules = []
follow_mouse_focus = True
bring_front_click = False
cursor_warp = False

floating_layout = layout.Floating(
    float_rules=[
        # Run the utility of `xprop` to see the wm class and name of an X client.
        *layout.Floating.default_float_rules,
        Match(wm_class="confirmreset"),  # gitk
        Match(wm_class="librewolf --profilemanager"),  # gitk
        Match(wm_class="makebranch"),  # gitk
        Match(wm_class="maketag"),  # gitk
        Match(wm_class="ssh-askpass"),  # ssh-askpass
        Match(title="branchdialog"),  # gitk
        Match(title="pinentry"),  # GPG key password entry
    ]
)

auto_fullscreen = False
focus_on_window_activation = "smart"
reconfigure_screens = True

# Pywal
# colors = []
# cache = os.path.expanduser( '~/.cache/wal/colors' )
# def load_colors(cache):
#    with open(cache, 'r') as file:
#        for i in range(8):
#            colors.append(file.readline().strip())
#    colors.append('#ffffff')
#    lazy.reload()
# load_colors(cache)


# Autostart
@hook.subscribe.startup
def autostart():
    home = os.path.expanduser("~/.config/qtile/scripts/autostart.sh")
    subprocess.Popen([home])


# If things like steam games want to auto-minimize themselves when losing
# focus, should we respect this or not?
auto_minimize = True

# Wayland
if qtile.core.name == "x11":
    term = "alacritty"
elif qtile.core.name == "wayland":
    term = "foot"

# XXX: Gasp! We're lying here. In fact, nobody really uses or cares about this
# string besides java UI toolkits; you can see several discussions on the
# mailing lists, GitHub issues, and other WM documentation that suggest setting
# this string if your java app doesn't work correctly. We may as well just lie
# and say that we're a working one by default.
#
# We choose LG3D to maximize irony: it is a 3D non-reparenting WM written in
# java that happens to be on java's whitelist.
wmname = "LG3D"
